### ISSUE TITLE
[Test] [YSQL] Table creation reports success before tservers finishes processing the request

### ISSUE THREAD

**Issue by frozenspider** [yugabyte/yugabyte-db#4056]
Here's the excerpt from logs of one of our Java tests:
```
m1|pid26116|:31694 I0320 12:24:35.162091 26183 ts_manager.cc:146] Registered new tablet server { permanent_uuid: "3d6501698e1b4bf69b57efbf02b26b53" instance_seqno: 1584707065973782 } with Master, full list: [{3d6501698e1b4bf69b57efbf02b26b53, 0x000000000433cfd0 -> { permanent_uuid: 3d6501698e1b4bf69b57efbf02b26b53 registration: common { private_rpc_addresses { host: "127.50.108.189" port: 19653 } http_addresses { host: "127.50.108.189" port: 29484 } cloud_info { placement_cloud: "cloud1" placement_region: "datacenter1" placement_zone: "rack1" } placement_uuid: "" } capabilities: 2189743739 placement_id: cloud1:datacenter1:rack1 }}, {5500304c92ea47a898488f0223938631, 0x000000000722ce10 -> { permanent_uuid: 5500304c92ea47a898488f0223938631 registration: common { private_rpc_addresses { host: "127.192.188.56" port: 18339 } http_addresses { host: "127.192.188.56" port: 28618 } cloud_info { placement_cloud: "cloud1" placement_region: "datacenter1" placement_zone: "rack1" } placement_uuid: "" } capabilities: 2189743739 placement_id: cloud1:datacenter1:rack1 }}, {b81bdbf9a1d74ea0983aacf784a27909, 0x000000000722c8d0 -> { permanent_uuid: b81bdbf9a1d74ea0983aacf784a27909 registration: common { private_rpc_addresses { host: "127.87.157.126" port: 12751 } http_addresses { host: "127.87.157.126" port: 32734 } cloud_info { placement_cloud: "cloud1" placement_region: "datacenter1" placement_zone: "rack1" } placement_uuid: "" } capabilities: 2189743739 placement_id: cloud1:datacenter1:rack1 }}]
m1|pid26116|:31694 I0320 12:24:35.162130 26183 catalog_manager.cc:740] 3 tablet servers registered, creating the transaction status table
m1|pid26116|:31694 I0320 12:24:35.162233 26183 catalog_manager.cc:2365] Creating the transaction status table
m1|pid26116|:31694 I0320 12:24:35.162348 26183 catalog_manager.cc:1818] CreateTable from internal request:
...
m1|pid26116|:31694 I0320 12:24:35.162494 26183 catalog_manager.cc:1963] Setting default tablets to 3 with 3 primary servers
m1|pid26116|:31694 I0320 12:24:35.162524 26183 partition.cc:388] Creating partitions with num_tablets: 3
m1|pid26116|:31694 I0320 12:24:35.162595 26183 catalog_manager.cc:2047] CreateTable with IndexInfo 
m2|pid26122|:11516 W0320 12:24:35.610877 26127 log.cc:706] Time spent Fsync log took a long time: real 0.447s	user 0.000s	sys 0.000s
m3|pid26129|:26840 W0320 12:24:35.664093 26134 log.cc:706] Time spent Fsync log took a long time: real 0.500s	user 0.000s	sys 0.000s
m1|pid26116|:31694 W0320 12:24:35.664110 26120 log.cc:706] Time spent Fsync log took a long time: real 0.501s	user 0.000s	sys 0.000s
m1|pid26116|:31694 W0320 12:24:36.190546 26247 long_operation_tracker.cc:112] TSHeartbeat running for 1.000s:
...
m1|pid26116|:31694 W0320 12:24:36.195783 26120 log.cc:706] Time spent Fsync log took a long time: real 0.531s	user 0.000s	sys 0.000s
...
m1|pid26116|:31694 I0320 12:24:36.779623 26183 catalog_manager.cc:2200] Successfully created table transactions [id=3213fc0e039946958c16ea8a4047b1b3] per request from internal request
m1|pid26116|:31694 I0320 12:24:36.779722 26183 catalog_manager.cc:753] Finished creating transaction status table asynchronously
m1|pid26116|:31694 I0320 12:24:36.779752 26183 catalog_manager.cc:3891] 3d6501698e1b4bf69b57efbf02b26b53 now has full report for 0 tablets.
m1|pid26116|:31694 I0320 12:24:36.779762 26183 catalog_manager.cc:3897] 3d6501698e1b4bf69b57efbf02b26b53 sent full tablet report with 0 tablets.
m1|pid26116|:31694 W0320 12:24:36.792671 26183 scoped_leader_shared_lock.cc:114] Long lock of catalog manager: 1.617s
...
2020-03-20 12:24:37,351 (Time-limited test) [INFO - org.yb.pgsql.BasePgSQLTest.initPostgresBefore(BasePgSQLTest.java:269)] Loading PostgreSQL JDBC driver
2020-03-20 12:24:37,356 (Time-limited test) [INFO - org.yb.pgsql.BasePgSQLTest.initPostgresBefore(BasePgSQLTest.java:275)] Waiting for initdb to complete on master
ts1|pid26190|:12751 I0320 12:24:37.357201 26228 tablet_service.cc:978] Processing CreateTablet for tablet ceca0ed16c884612a20298aba210c830 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [43690, <end>)
ts1|pid26190|:12751 I0320 12:24:37.357383 26228 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ceca0ed16c884612a20298aba210c830
ts1|pid26190|:12751 I0320 12:24:37.360183 26225 tablet_service.cc:978] Processing CreateTablet for tablet ca963657dae74b77bf35fc10033c5a0a (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [21845, 43690)
ts1|pid26190|:12751 I0320 12:24:37.360219 26225 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ca963657dae74b77bf35fc10033c5a0a
ts2|pid26229|:19653 I0320 12:24:37.358530 26265 tablet_service.cc:978] Processing CreateTablet for tablet ceca0ed16c884612a20298aba210c830 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [43690, <end>)
ts2|pid26229|:19653 I0320 12:24:37.358697 26265 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ceca0ed16c884612a20298aba210c830
ts2|pid26229|:19653 I0320 12:24:37.359500 26262 tablet_service.cc:978] Processing CreateTablet for tablet bee8bbc499d140c197b91809a37d8a37 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [<start>, 21845)
ts2|pid26229|:19653 I0320 12:24:37.359526 26262 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet bee8bbc499d140c197b91809a37d8a37
ts2|pid26229|:19653 I0320 12:24:37.361807 26266 tablet_service.cc:978] Processing CreateTablet for tablet ca963657dae74b77bf35fc10033c5a0a (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [21845, 43690)
ts2|pid26229|:19653 I0320 12:24:37.361838 26266 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ca963657dae74b77bf35fc10033c5a0a
ts3|pid26279|:18339 I0320 12:24:37.363909 26306 tablet_service.cc:978] Processing CreateTablet for tablet ca963657dae74b77bf35fc10033c5a0a (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [21845, 43690)
ts3|pid26279|:18339 I0320 12:24:37.364050 26306 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ca963657dae74b77bf35fc10033c5a0a
ts3|pid26279|:18339 I0320 12:24:37.364132 26302 tablet_service.cc:978] Processing CreateTablet for tablet bee8bbc499d140c197b91809a37d8a37 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [<start>, 21845)
ts3|pid26279|:18339 I0320 12:24:37.364177 26302 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet bee8bbc499d140c197b91809a37d8a37
ts3|pid26279|:18339 I0320 12:24:37.364418 26309 tablet_service.cc:978] Processing CreateTablet for tablet ceca0ed16c884612a20298aba210c830 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [43690, <end>)
ts3|pid26279|:18339 I0320 12:24:37.364446 26309 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet ceca0ed16c884612a20298aba210c830
ts1|pid26190|:12751 I0320 12:24:37.363842 26227 tablet_service.cc:978] Processing CreateTablet for tablet bee8bbc499d140c197b91809a37d8a37 (table=transactions [id=3213fc0e039946958c16ea8a4047b1b3]), partition=hash_split: [<start>, 21845)
ts1|pid26190|:12751 I0320 12:24:37.363873 26227 ts_tablet_manager.cc:1667] Get and update data/wal directory assignment map for table: 3213fc0e039946958c16ea8a4047b1b3 and tablet bee8bbc499d140c197b91809a37d8a37
m1|pid26116|:31694 W0320 12:24:37.366855 26200 scoped_leader_shared_lock.cc:114] Long lock of catalog manager: 1.782s
...
2020-03-20 12:24:37,370 (Time-limited test) [INFO - org.yb.pgsql.BasePgSQLTest.initPostgresBefore(BasePgSQLTest.java:290)] initdb has completed successfully on master
ts1|pid26190|:12751 I0320 12:24:37.409524 26381 ybc_util.cc:62] Process info is written to /tmp/yb_test.tmp.754.1843.23493.pid25995/process_info/26381
ts1|pid26190|:12751 I0320 12:24:37.410430 26381 pggate.cc:106] Reset YSQL bind address to 127.87.157.126:5432
...
ts1|pid26190|:12751 I0320 12:24:37.417111 26388 async_initializer.cc:77] Successfully built ybclient
ts1|pid26190|:12751 2020-03-20 12:24:37.850 UTC [26381] ERROR:  Illegal state: No tablets in table system.transactions: . Errors from tablet servers: [Illegal state (yb/client/transaction_manager.cc:88): No tablets in table system.transactions]
ts1|pid26190|:12751 2020-03-20 12:24:37.850 UTC [26381] STATEMENT:  CREATE ROLE yugabyte_test SUPERUSER CREATEROLE CREATEDB BYPASSRLS LOGIN
```
(full test log: [initial-test.log.zip](https://github.com/yugabyte/yugabyte-db/files/4375801/initial-test.log.zip))
From here it appears that the request to create a transaction status table (essential for YSQL) has been processed by master and forwarded to tservers, but the next request arrived before the table was actually created by tservers.
I was able to simulate this by adding a delay to `TabletServiceAdminImpl::CreateTablet`

**Comment by bmatican** [comment]
@frozenspider I don't recall what we discussed last time, if we wanted to put in the wait for txn tablet in `ScopedLeaderLock` or in `GetTableLocations`, if executed for the txn table?

**Comment by frozenspider** [comment]
@bmatican We discussed both options, and while the second one is simpler, we decided that the first one would probably be more appropriate place for this. I only had time to verify that the second one works though, before switching to another issue.

**Comment by bmatican** [comment]
Posting some clarification.

In `CatalogManager::VisitSysCatalog`, we have some code to create the `transactions` table, if needed (loop until `OK` or `AlreadyPresent`). However, our create table calls are async, and normally client driven, such that clients can also wait for them to be complete. In this case, there is no wait being performed!

Moreover, we have to wait until we have a minimum of servers, before we can even start to create this table, otherwise the create will fail validation. So really, the sequence should be:
- the master needs to become leader
- finish loading metadata
- wait for heartbeats from at least 3 servers (RF3)
- then we async kick off the create of this table
- (MISSING) someone needs to wait on this table being created
- then we can get client requests

Currently, we're failing in SQL tests because the local TS side code checks if there are any transaction status tablets, which it does using `GetTableLocations`. So our initial though was to wire that up to wait for the table, if it's asking specifically about this table. I believe this is what @frozenspider already validated works and fixes the tests.

However, there might be other requests which require the transactions table to exist. Or in the future, maybe the ordering of calls from the TS might change? So a more generic location to put such a wait could he the top-level `SharedLeaderLock`, which all requestst that need to be executed on the master leader, must take.

cc @ndeodhar @mbautin
